File: solutions/python/meetup/1/meetup.py
from calendar import Calendar
from datetime import date


# subclassing the built-in ValueError to create MeetupDayException
class MeetupDayException(ValueError):
    """Exception raised when the Meetup weekday and count do not result in a valid date.

    message: explanation of the error.
    """


def select_day(selector: str, days: list[int]) -> int:
    try:
        return {
            "first": lambda: days[0],
            "second": lambda: days[1],
            "third": lambda: days[2],
            "fourth": lambda: days[3],
            "fifth": lambda: days[4],
            "last": lambda: days[-1],
            "teenth": lambda: [d for d in days if 10 <= d < 20][-1],
        }[selector]()
    except IndexError as exc:
        raise MeetupDayException("That day does not exist.") from exc


def meetup(year: int, month: int, selector: str, day_of_week: str) -> date:
    target_weekday = {
        "Monday": 0,
        "Tuesday": 1,
        "Wednesday": 2,
        "Thursday": 3,
        "Friday": 4,
        "Saturday": 5,
        "Sunday": 6,
    }[day_of_week]
    days = [
        monthday
        for monthday, weekday in Calendar().itermonthdays2(year, month)
        if monthday and weekday == target_weekday
    ]
    day = select_day(selector, days)
    return date(year, month, day)

File: solutions/go/meetup/1/meetup.go
package meetup

import (
	"time"
)

type WeekSchedule string

var (
	First  WeekSchedule = "First"
	Second WeekSchedule = "Second"
	Third  WeekSchedule = "Third"
	Fourth WeekSchedule = "Fourth"
	Teenth WeekSchedule = "Teenth"
	Last   WeekSchedule = "Last"
)

func Day(
	wSched WeekSchedule, wDay time.Weekday, month time.Month, year int,
) int {
	var last time.Time
	start := map[WeekSchedule]int{
		First: 1, Second: 8, Third: 15, Fourth: 22, Teenth: 13,
		Last: 22,
	}[wSched]
	for day := start; ; day++ {
		date := time.Date(year, month, day, 0, 0, 0, 0, time.UTC)
		if date.Month() != month {
			break
		}
		if date.Weekday() != wDay {
			continue
		}
		if wSched != Last {
			return date.Day()
		}
		last = date
	}
	return last.Day()
}
